kvs: refactor with internal kvsroot file

chu11 opened the pull request:
This branch isn't ready for merging (need to add documentation & unit tests) but thought I'd push to ask for comments.

I had originally gone down the path of abstracting the kvsroot way more than what is done in this PR.  But as I was refactoring, it became clear abstracting ```struct kvsroot``` was creating a complex and probably unnecessary level of abstraction.

For example, abstracting ```start_root_remove()``` in ```kvs.c```.  It's a good idea to want to abstract it away, but its difficult to abstract away the cleanup of fences sitting in the commit manager.  Do we ask the user to pass in a callback function to deal with the cleanup of fences?  We aren't abstracting very well if we're asking the user for that.

I briefly considered moving everything related to commits into ```kvsroot.c```, but seemed like too much.  So I think other refactoring needs to be done before moving onto refactoring kvsroot more.  Some thoughts for the future.

- rethink how fences are cleaned up in its entirety (i.e. lookup stored msg copies based on name maybe not best).  Perhaps some callback mechanism would work (i.e. when fence deleted, callback something to complete it.)
- rethink entire abstraction of "fences" in ```fence.[ch]```.  Perhaps everything should be collapsed into the commit manager in some way.

garlick commented:
Looks like a good direction.

I restarted a travis build that had stalled for no apparent reason.

chu11 commented:
ok, let me add documentation & unit tests, then this should be all good to go.

chu11 commented:
added a few header comments in ```kvsroot.h``` where I thought they were necessary.  Then added unit tests.  I already squashed all changes into the original set as there weren't too many differences with the comment header changes.

garlick commented:
Looks good @chu11 - no coverage report here (?) but it was +0.003% in the last report and it seems it can't help but have gone up from there with the new tests, so I'm good to merge if you're ready.

chu11 commented:
yup, all done.
